lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -6,7 +6,14 @@ defmodule Mix.Compilers.Elixir do
   import Record
 
   defrecordp :module, [:module, :kind, :source, :beam, :binary]
-  defrecordp :source, [:source, compile: [], runtime: [], external: []]
+  defrecordp :source, [
+    source: nil,
+    compile: [],
+    runtime: [],
+    compile_dispatches: [],
+    runtime_dispatches: [],
```

Comment:
Ok, after some more thought, I am thinking we should ignore compile_dispatches and simply store runtime_dispatches. We could change this up in the lexical tracker itself and call "dispatches" simply everything that will happen at runtime. The rationale is the following: all the compile time dispatches will be invoked at compile time anyway. The only exception is when the code is not meant to be invoked at all which means we should not validate it as well.

For example, imagine this code at the module level, outside of a function:

```elixir
if function_exported?(List, :flatten, 1) do
  List.flatten([1, 2, 3])
else
  List.old_flatten([1, 2, 3])
end
```

It is common code that checks for a particular version and acts accordingly. We could even add a test that ensures the code above won't emit any warnings.

 At the function body, there is nothing we can do, but we can at least define internal functions that act differently:

```elixir
if function_exported?(List, :flatten, 1)
  def flatten(arg), do: List.flatten(arg)
else
  def flatten(arg), do: List.old_flatten(arg)
end
```

And because only one function will be defined, the runtime will see only one of the two versions and not emit warnings.